spring-cloud-stream-reactive/src/main/java/org/springframework/cloud/stream/reactive/StreamEmitterAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,273 @@
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link BeanPostProcessor} that handles {@link StreamEmitter} annotations found on bean methods.
+ *
+ * @author Soby Chacko
+ * @since 1.3.0
+ */
+public class StreamEmitterAnnotationBeanPostProcessor
+		implements BeanPostProcessor, InitializingBean, ApplicationContextAware, SmartLifecycle {
+
+	private final List<StreamListenerParameterAdapter<?, Object>> streamListenerParameterAdapters = new ArrayList<>();
+
+	private final List<StreamListenerResultAdapter<?, ?, ?>> streamListenerResultAdapters = new ArrayList<>();
+
+	private final List<FluxSender> fluxSenders = new ArrayList<>();
+
+	private final List<Disposable> fluxDisposables = new ArrayList<>();
+
+	private ConfigurableApplicationContext applicationContext;
+
+	private MultiValueMap<Object, Method> mappedStreamEmitterMethods = new LinkedMultiValueMap<>();
+
+	private volatile boolean running;
+
+	private Lock lock = new ReentrantLock();
+
+	@Override
+	public final void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		Assert.isTrue(applicationContext instanceof ConfigurableApplicationContext,
+				"ConfigurableApplicationContext is required");
+		this.applicationContext = (ConfigurableApplicationContext) applicationContext;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void afterPropertiesSet() throws Exception {
+		Map<String, StreamListenerParameterAdapter> parameterAdapterMap = BeanFactoryUtils
+				.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerParameterAdapter.class);
+		parameterAdapterMap.values().iterator().forEachRemaining(this.streamListenerParameterAdapters::add);
+		Map<String, StreamListenerResultAdapter> resultAdapterMap = BeanFactoryUtils
+				.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerResultAdapter.class);
+		this.streamListenerResultAdapters.add(new MessageChannelStreamListenerResultAdapter());
+		resultAdapterMap.values().iterator().forEachRemaining(this.streamListenerResultAdapters::add);
+	}
+
+	@Override
+	public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+		return bean;
+	}
+
+	@Override
+	public Object postProcessAfterInitialization(final Object bean, final String beanName) throws BeansException {
+		Class<?> targetClass = AopUtils.isAopProxy(bean) ? AopUtils.getTargetClass(bean) : bean.getClass();
+		ReflectionUtils.doWithMethods(targetClass, method -> {
+			StreamEmitter streamEmitter = AnnotatedElementUtils.findMergedAnnotation(method, StreamEmitter.class);
+			if (streamEmitter != null && !method.isBridge()) {
+				mappedStreamEmitterMethods.add(bean, method);
+			}
+		});
+		return bean;
+	}
+
+	@Override
+	public void start() {
+		try {
+			lock.lock();
+			if (!running) {
+				mappedStreamEmitterMethods.forEach((k, v) -> v.forEach(item -> {
```

Comment:
Hm. Seems for me `MultiValueMap` is an overhead here.
Why simple `Map<Method, Object>` isn't enough ?